integrations/influxdb_metrics_dashboard/lambda/upload_dashboard/main.go

```diff
@@ -0,0 +1,459 @@
+	req, err := http.NewRequest(requestType, "https://"+urlWithEndpoint, payload)
+	if err != nil {
+		log.Printf("Failed to create Grafana HTTP request: %s", err)
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Authorization", "Bearer "+serviceAccountTokenKey)
+	return req, nil
+}
+
+func sendGrafanaHttpReq(httpClient http.Client, req *http.Request) (*http.Response, error) {
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		log.Printf("Failed to execute Grafana HTTP request: %s", err)
+		return nil, err
+	}
+	return resp, nil
+}
+
+func uploadDashboard(serviceAccountTokenKey string, workspaceUrl string, datasourceName string, dashboardName string, databaseName string) (string, error) {
+	const SleepDuration = 10
+	const MaxWaitIntervals = 100
+
+	var plugins []struct {
+		Id   string `json:"id"`
+		Name string `json:"name"`
+	}
+
+	httpClient := &http.Client{
+		Timeout: 30 * time.Second,
+	}
+
+	installTimestreamPluginReq, err := getGrafanaHttpReq(
+		"POST",
+		workspaceUrl+"/api/plugins/grafana-timestream-datasource/install",
+		nil,
+		serviceAccountTokenKey,
+	)
+	if err != nil {
+		return "", err
+	}
+	installTimestreamPluginResp, err := sendGrafanaHttpReq(*httpClient, installTimestreamPluginReq)
+	if err != nil {
+		log.Printf("Failed to install Timestream plugin in Grafana workspace: %s", err)
+		return "", err
+	}
+
+	if installTimestreamPluginResp.StatusCode == http.StatusConflict {
+		log.Printf("Timestream plugin is already installed in the Grafana workspace")
+	} else if installTimestreamPluginResp.StatusCode != http.StatusOK {
+		log.Printf("Received status code %d when trying to install plugin: %s", installTimestreamPluginResp.StatusCode, err)
+		return "", fmt.Errorf("error: %d", installTimestreamPluginResp.StatusCode)
+	}
+	log.Printf("Timestream data source installed")
+
+	getGrafanaPluginsReq, err := getGrafanaHttpReq(
+		"GET",
+		workspaceUrl+"/api/plugins",
+		nil,
+		serviceAccountTokenKey,
+	)
+	if err != nil {
+		return "", err
+	}
+
+	getGrafanaPluginsResp, err := sendGrafanaHttpReq(*httpClient, getGrafanaPluginsReq)
+	if err != nil {
+		log.Printf("Failed to retrieve plugins for Grafana workspace: %s", err)
+		return "", err
+	}
+
+	waiterInterval := 0
+	for waiterInterval < MaxWaitIntervals {
+		// Let Grafana catch up
+		time.Sleep(SleepDuration * time.Second)
+		body, err := io.ReadAll(getGrafanaPluginsResp.Body)
+		if err != nil {
+			log.Printf("Failed to read response body %s", err)
+			return "", err
+		}
+		if err := json.Unmarshal(body, &plugins); err != nil {
+			log.Printf("Failed to unmarshal JSON %s", err)
+			return "", err
+		}
+		pluginInstalled := false
+		for _, plugin := range plugins {
+			if plugin.Name == "Amazon Timestream" {
+				pluginInstalled = true
+				break
+			} else {
```

Comment:
Wouldn't this always check just the first returned plugin for the workspace, and not loop through any of the other returned plugins, and if the first plugin isn't the Timestream plugin, then make the `getGrafanaPluginsReq` again?